src/memcpy_vfp.c

```diff
@@ -0,0 +1,151 @@
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "driver.h"
+
+#define BLOCK_SIZE  4096
+
+static char tmp_buf0[BLOCK_SIZE] __attribute__((aligned (128)));
+
+static inline void tegra_aligned_block_copy_vfp(void *dst, void *src, int size)
+{
+    asm volatile(
+        "   .fpu vfpv3-d16          \n\t"
+        "   .arch armv7a            \n\t"
+        "0:                         \n\t"
+        "   pld   [%1, #128]        \n\t"
+        "   pld   [%1, #160]        \n\t"
+        "   pld   [%1, #192]        \n\t"
+        "   pld   [%1, #224]        \n\t"
+        "   pld   [%1, #256]        \n\t"
+        "   vldm  %1!, {d0-d15}     \n\t"
+        "   subs  %2, %2, #128      \n\t"
+        "   vstm  %0!, {d0-d15}     \n\t"
+        "   bgt   0b                \n\t"
+        : "+r" (dst), "+r" (src), "+r" (size)
+        :
+        : "cc", "d8", "d9", "d10", "d11", "d12", "d13", "d14", "d15");
+}
+
+void tegra_copy_block_vfp(char *dst, char *src, int size,
+                          bool bounce, bool move)
+{
+    int i, dir, block_size = BLOCK_SIZE;
+    char *tmpbuf = tmp_buf0;
+    char *psrc = src;
+    char *pdst = dst;
+
+    assert(((uintptr_t)src & 127) == 0);
+    assert(((uintptr_t)dst & 127) == 0);
+    assert((size & 127) == 0);
+
+    do {
+        if (size < block_size || !bounce) {
+            block_size = size;
+            move = false;
+        }
+
+        dir = (pdst > psrc && move) ? -1 : 1;
+
+        if (dir < 0) {
+            psrc += size - block_size;
+            pdst += size - block_size;
+        }
+
+        for (i = 0; i < size / block_size; i++) {
+            if (!bounce)
+                tmpbuf = pdst;
+
+            tegra_aligned_block_copy_vfp(tmpbuf, psrc, block_size);
+
+            if (bounce)
+                tegra_aligned_block_copy_vfp(pdst, tmpbuf, block_size);
+
+            psrc += block_size * dir;
+            pdst += block_size * dir;
+        }
+
+        size -= block_size * i;
+        pdst = max(pdst, dst);
+        psrc = max(psrc, src);
+
+    } while (size);
+}
+
+void tegra_do_memcpy_vfp(void *dst, const void *src, int size, bool bounce)
+{
+    bool src_aligned = !((uintptr_t)src & 127);
+    bool dst_aligned = !((uintptr_t)dst & 127);
+    int left = size;
+    int block_size;
+    char *csrc = (char *)src;
```

Comment:
Why not `const char *`?

Reply:
The code was changing and this seems can be changed now to `const char *`, thanks.